src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java

```diff
@@ -0,0 +1,193 @@
+        return size(segments);
+    }
+
+
+    @Override
+    public CompletionStage<Boolean> containsKey(int segment, Object key) {
+        return load(segment, key).thenApply(Objects::nonNull);
+    }
+
+    @Override
+    public CompletionStage<MarshallableEntry<K, V>> load(int segment, Object key) {
+        long now = context.getTimeService().wallClockTime();
+        Object storeKey = cacheToStoreConverter.toStoreKey(key);
+        Bson filter = and(eq("_id", storeKey), or(gt(EXPIRY_TIME, now), lte(EXPIRY_TIME, -1L)));
+
+        return Maybe.fromPublisher(collection.find(filter))
+                .map(document -> cacheToStoreConverter.toCacheEntry(key, document))
+                .toCompletionStage(null);
+    }
+
+    @Override
+    public CompletionStage<Void> write(int segment, MarshallableEntry<? extends K, ? extends V> marshallableEntry) {
+        Object storeKey = cacheToStoreConverter.toStoreKey(marshallableEntry.getKey());
+        Document storeEntry = cacheToStoreConverter.toStoreEntry(storeKey, marshallableEntry);
+
+        Publisher<UpdateResult> publisher = collection.replaceOne(eq("_id", storeKey), storeEntry, new ReplaceOptions().upsert(true));
+        return Single.fromPublisher(publisher)
+                .ignoreElement()
+                .toCompletionStage(null);
+    }
+
+    @Override
+    public CompletionStage<Boolean> delete(int segment, Object key) {
+        BasicDBObject query = new BasicDBObject();
+        query.put("_id", cacheToStoreConverter.toStoreKey(key));
+
+        return Single.fromPublisher(collection.deleteOne(query))
+                .map(e -> e.getDeletedCount() > 0)
+                .toCompletionStage();
+    }
+
+    @Override
+    public CompletionStage<Void> clear() {
+        return Completable.fromPublisher(collection.drop())
+                .toCompletionStage(null);
+    }
+
+
+    @Override
+    public Publisher<MarshallableEntry<K, V>> publishEntries(IntSet segments, Predicate<? super K> filter, boolean includeValues) {
+        return Flowable.defer(() -> {
+            FindPublisher<Document> iterable = collection.find();
+
+            if (!includeValues) {
+                iterable.projection(Projections.include("_id"));
+            }
+
+            return Flowable.fromPublisher(iterable)
+                    .filter(e -> filter == null || filter.test(cacheToStoreConverter.toCacheKey(e.get("_id"))))
+                    .map(cacheToStoreConverter::toCacheEntry);
+        });
+    }
+
+    @Override
+    public Publisher<MarshallableEntry<K, V>> purgeExpired() {
+        long now = context.getTimeService().wallClockTime();
+        Bson filter = and(lte(EXPIRY_TIME, now), gt(EXPIRY_TIME, -1L));
+
+        return Flowable.fromPublisher(collection.find(filter))
+                .map(cacheToStoreConverter::toCacheEntry)
+                .concatWith(Flowable.defer(() -> Flowable.fromPublisher(collection.deleteMany(filter)).mapOptional(x -> Optional.empty())));
```

Comment:
Unfortunately, this has a pretty big window where delete may not return the entries that were actually deleted.

Looking at the mongo API there isn't a great way to correlate a delete to a specific key. I wonder though if bulkWrite with a List of DeleteOneModel instances will return a Publisher with the same results if the option is ordered. If so then we can return a consistent view of what entries were actually removed due to expiration.

Reply:
Something like this?
```java
        return Flowable.fromPublisher(collection.find(filter))
                .buffer(context.getConfiguration().maxBatchSize())
                .concatMap(list -> {
                    var deleteModels = list.stream()
                            .map(e -> new DeleteOneModel<Document>(eq("_id", e.get("_id"))))
                            .toList();
                    return Flowable.fromPublisher(collection.bulkWrite(deleteModels))
                            .mapOptional(ignored -> Optional.<MarshallableEntry<K, V>>empty())
                            .concatWith(Flowable.fromStream(list.stream().map(cacheToStoreConverter::toCacheEntry)));
                });
```

In `BulkWriteResult` upserts don't take into account deletes, so I don't see a way to get the ids/documents back.

Reply:
This is very close, we should also add the expiration time to the DeleteOneModel so it only deletes the entry if it is still expired. It is possible for a concurrent update that could change the expiration time.

I think the publishing of the entry based on if it was expired isn't quite right either

I believe the following should work, assuming the BulkWriteResult are returned in the same order as the write

```java
        return Flowable.fromPublisher(collection.find())
              .buffer(context.getConfiguration().maxBatchSize())
              .concatMap(list -> {
                  var deleteModels = list.stream()
                        .map(e -> new DeleteOneModel<Document>(eq("_id", e.get("_id"))), <expiration here>)
                        .toList();
                  return Flowable.zip(Flowable.fromPublisher(collection.bulkWrite(deleteModels)),
                        Flowable.fromIterable(list),
                        (bwr, entry) -> {
                            if (bwr.wasAcknowledged() && bwr.getDeletedCount() > 0) {
                                return Optional.of(cacheToStoreConverter.toCacheEntry(entry));
                            }
                            return Optional.<MarshallableEntry<K, V>>empty();
                        })
                        .mapOptional(Functions.identity());
              });
```

Reply:
Unfortunately `bulkWrite` returns a publisher with a single result.

Reply:
That is quite disappointing.... Without that, we won't have proper expiration notifications for entries that are in the store but not in memory. The store will just need to be documented with that limitation.

Reply:
What do you think about this?

```java
    public Publisher<MarshallableEntry<K, V>> purgeExpired() {
        long now = context.getTimeService().wallClockTime();

        return Flowable.fromPublisher(collection.find(and(lte(EXPIRY_TIME, now), gt(EXPIRY_TIME, -1L))).projection(Projections.include("_id")))
                .concatMapMaybe(d -> Single.fromPublisher(collection.findOneAndDelete(and(eq("_id", d.get("_id")), lte(EXPIRY_TIME, now), gt(EXPIRY_TIME, -1L))))
                        .mapOptional(docOrNull -> Optional.ofNullable(docOrNull).map(doc -> cacheToStoreConverter.toCacheEntry(doc))));
    }
```

Of course it's not the most efficient (without batching), but at least it doesn't iterate over the entire collection and should be safe.

Also it's not optimal, since it may miss some expired entries: is it a problem?

Reply:
I am fine either way. This should be fine the way it is, the entries can just be found on the next pass.

The only thing that worries me is that the `MongoCollection` API seems to be breaking reactive streams specification with `If no documents matched the query filter, then null will be returned`. A publisher cannot contain a null value, so I am not sure this will work or not...

Reply:
Sounds good.

Reply:
If you want to update to either do the more efficient removal but we don't have notifications or we can do the slower but have notifications. I guess I would lean towards the latter for correctness, but I don't know how much this will affect the mongo server with all those requests.

Reply:
Same here.